### PR TITLE
Shipping methods extension hook

### PR DESCRIPTION
**Pull request by sauloperez**
## What

This is a first attempt at providing extensions hooks of our own to Spree similar to what Solidus does. It is a first step towards solving https://github.com/openfoodfoundation/openfoodnetwork/issues/2205 in a more maintainable way.

Doing this gives us the possibility to provide our own implementation of `Package#shipping_methods`, based on Spree's one, and instead of totally overriding it in a `class_eval` block.

## Missing

- [x] :warning: requires https://github.com/coopdevs/spree/pull/5 being merged first and then rebase onto it.
- [x] Point out  in the README the fact that we only add extension hooks and not customizations to this fork

### CODE REVIEW

File: core/spec/models/spree/stock/packer_spec.rb [hunk 1]
```diff
@@ -31,6 +35,39 @@ module Stock
...
+        end
+
+        context 'when a packer factory is specified' do
+          before do
+            Spree::Config.stub(:package_factory) { TestPackageFactory }
+          end
+
+          class TestPackageFactory; end
+
+          let(:package_factory) { TestPackageFactory }
```
Comment (sauloperez) [inline core/spec/models/spree/stock/packer_spec.rb]: Partially used :warning:

File: core/app/models/spree/stock/packer.rb [hunk 2]
```diff
@@ -1,3 +1,5 @@
+require 'byebug'
+
```
Comment (enricostano) [inline core/app/models/spree/stock/packer.rb]: :eyes:

**Comment by enricostano** [comment]
Cool! My only question is if we should consider adding this "hooks layer" in a separate gem instead of forking Spree so radically. I'm just concerned about upgrading Spree in next iterations with all these hooks added directly into Spree code.

File: core/app/models/spree/stock/packer.rb (hunk 2 above)
Comment (sauloperez) [reply core/app/models/spree/stock/packer.rb]: lol

**Comment by sauloperez** [comment]
what do you have in mind when you say in a separate gem? The issue here is that it's Spree's code base which has to give support for this extension points; calling the configured package class rather than the Spree's built in one.

> I'm just concerned about upgrading Spree in next iterations with all these hooks added directly into Spree code.

Me too. By doing things explicitly, however, we benefit from the test suite. Once we apply the hooks on top of a new Spree version the specs we added we'll cover us. The failing specs we'll tell us which hooks need to be refactored to fit Spree's API changes. No more guessing or noticing these changes through failures in production.

The way I can think of doing this backport is grabbing the bunch of commits that implement the hooks and cherrypicking them on top of the next Spree version. Do you see a better way?

**Pull request merged**